R/dependencies.R

```diff
@@ -0,0 +1,410 @@
+    if (circ) {
+      tree$message <- "There appears to be a circular dependency."
+      return(tree)
+    }
+  }
+
+  # make sure a report with this name exists
+  db_reports <- DBI::dbGetQuery(con, "SELECT name FROM report")
+  if (!(name %in% db_reports$name)) {
+    stop("This report does not exist")
+  }
+
+  # do we need to find the latest version of the report?
+  if (id == "latest") {
+    id <- get_latest_by_name(con, name)$id
+  } else {
+    db_id <- id_to_name(con, id)
+    if (is.null(db_id)) {
+      stop(sprintf("No report with id %s in the database", id))
+    } else {
+      if (name != id_to_name(con, id)) {
+        stop(sprintf("id %s does not match report name %s", id, name))
+      }
+    }
+  }
+
+  if (upstream) {
+    ## when going upstream the concept of out-of-date doesn't make sense?
+    ood <- is_out_of_date(con, id)
+  } else {
+    ood <- is_out_of_date(con, id)
+  }
+
+  ## if this is no tree, create a tree...
+  if (is.null(tree)) {
+    v <- Vertex$new(NULL, name, id, ood)
+    tree <- Tree$new(v)
+  } else { ## ...otherwise add a vertex
+    v <- tree$add_child(parent, name, id, ood)
+  }
+
+  dep_ids <- get_dependencies_db(name = name, id = id, con = con,
+                                 upstream = upstream, list_all = list_all)
+  for (dep_id in dep_ids) {
+    dep_name <- id_to_name(id = dep_id, con = con)
+    build_tree(name = dep_name, id = dep_id, depth = depth - 1, parent = v,
+               tree = tree, con = con, upstream = upstream, list_all = list_all)
+  }
+
+  return(tree)
+}
+
+##' @title Given a tree return a list of reports to be re-run (and the order
+##' that they should be re-run)
+##'
+##' @tree A vertex object
+##'
+##' @return A list of report names
+out_ot_date_reports <- function(vertex, reports = c()) {
+  if (vertex$out_of_date) {
+    if (vertex$name %in% reports) {
+      reports <- setdiff(reports, vertex$name)
+    }
+    reports <- c(reports, vertex$name) ## add it to the list
+  }
+
+  if (length(vertex$children) > 0) {
+    for (i in 1:length(vertex$children)) {
+      child <- vertex$children[[i]]
+
+      reports <- out_ot_date_reports(child, reports)
+    }
+  }
+  reports
+}
+
+##' @title Print the dependency tree for a given report using orderly log
+##'
+##' @param name the name of the report
+##' @param id the id of the report, if omitted, use the id of the latest report
+##' @param con A connection to a database
+##' @param root
+##' @param locate
+##' @param upstream A boolean indicating if we want to move up or down the tree
+##' @param propagate A boolean indicating if we want to propogate out of date
+##'                  through the tree
+##'
+##' @export
+orderly_build_dep_tree <- function(name, id = "latest", root = NULL,
+                                   locate = TRUE, upstream = FALSE, con = NULL,
+                                   propagate = FALSE, max_depth = 100,
+                                   list_all = FALSE) {
+  if (is.null(con)) {
```

Comment:
I don't see a use case for accepting both root/locate and con - drop the con argument and make the connection unconditional